src/repr/type_binary.ml

```diff
@@ -244,8 +241,79 @@ module Decode = struct
         unstage decoders.(i) buf ofs)
 end
 
+module Pre_hash = struct
+  type 'a pre_hash = 'a encode_bin
+
+  let rec t : type a. a t -> a pre_hash = function
+    | Self s -> self s
+    | Custom c -> c.pre_hash
+    | Map m -> map m
+    | Boxed b -> t b
+    | Attributes { attr_type; _ } -> t attr_type
+    | List l -> Encode.list (t l.v) l.len
+    | Array a -> Encode.array (t a.v) a.len
+    | Tuple t -> tuple t
+    | Option x -> Encode.option (t x)
+    | Record r -> record r
+    | Variant v -> variant v
+    | Var v -> raise (Unbound_type_variable v)
+    | Prim _ as t -> Encode.t t
+
+  and self : type a. a self -> a pre_hash =
+   fun { self_unroll; _ } ->
+    fix_staged (fun pre_hash ->
+        let cyclic = self_unroll (partial ~pre_hash ()) in
+        t cyclic)
+
+  and tuple : type a. a tuple -> a pre_hash = function
+    | Pair (x, y) -> Encode.pair (t x) (t y)
+    | Triple (x, y, z) -> Encode.triple (t x) (t y) (t z)
+
+  and map : type a b. (a, b) map -> b pre_hash =
+   fun { x; g; _ } ->
+    let pre_hash = unstage (t x) in
+    stage (fun u f -> pre_hash (g u) f)
+
+  and record : type a. a record -> a encode_bin =
+   fun r ->
+    let field_encoders : (a -> (string -> unit) -> unit) list =
+      ListLabels.map (fields r) ~f:(fun (Field f) ->
+          let field_encode = unstage (t f.ftype) in
+          fun x -> field_encode (f.fget x))
+    in
+    stage (fun x k -> Stdlib.List.iter (fun f -> f x k) field_encoders)
+
+  and variant : type a. a variant -> a encode_bin =
+    let c0 { ctag0; _ } = stage (Bin.Int.encode ctag0) in
+    let c1 c =
+      let encode_arg = unstage (t c.ctype1) in
+      stage (fun v k ->
+          Bin.Int.encode c.ctag1 k;
+          encode_arg v k)
+    in
+    fun v -> fold_variant { c0; c1 } v
+
+  (* NOTE: for compatibility reasons, the pre-hash of a primitive type uses the
+     _unboxed_ encoding of that type (but primitive components of larger types
+     _are_ boxed). *)
+  let t =
```

Comment:
This solution has the downside of having two different behaviours for these 2 examples:
```
type elt = string [@@deriving repr]

type string_pair = elt * elt [@@deriving repr]
```

```
type elt = string [@@deriving repr]

let elt_t = Repr.(like ~pre_hash:(pre_hash elt_t) elt_t)

type string_pair = elt * elt [@@deriving repr]
```

(Not that it matters much)

Reply:
Agreed. I'd have just boxed everything unconditionally if not for backwards compatibility concerns.